File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.js
function solution(arr) {
    const N = arr.length;
    const answer = [0, 0]; // [0 개수, 1 개수]
    
    const dfs = (size, r, c) => {
        const first = arr[r][c];
        if(size === 1) {
            answer[first]++;
            return;
        }
        
        let isUniform = true;
        for(let i = r; i < r + size; i++) {
            for(let j = c; j < c + size; j++) {
                if(arr[i][j] !== first) {
                    isUniform = false;
                    break;
                }
            }
            if(!isUniform) break;
        }
        
        if(isUniform) {
            answer[first]++;
            return;
        } else {
            const half = size / 2;
            dfs(half, r, c);
            dfs(half, r + half, c);
            dfs(half, r, c + half);
            dfs(half, r + half, c + half);
        }
    }
    
    dfs(N, 0, 0);
    return answer;
}

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
def solution(arr):
    n = len(arr)
    answer = [0, 0] # [0 개수, 1 개수]
    def dfs(size, r, c):
        first = arr[r][c]
        
        if size == 1:
            answer[first] += 1
            return
        
        is_uniform = True
        for i in range(r, r + size):
            for j in range(c, c + size):
                if arr[i][j] != first:
                    is_uniform = False
                    break
            if not is_uniform:
                break
        
        if is_uniform:
            answer[first] += 1
        else:
            half = size // 2
            dfs(half, r, c)
            dfs(half, r + half, c)
            dfs(half, r, c + half)
            dfs(half, r + half, c + half)
    
    dfs(n, 0, 0)
    return answer